File: pkg/tracex/topology.go
package tracex

// TopologyGraph is the static component graph served to the dashboard. It
// mirrors the ARCHITECTURE.md mermaid diagram and the channel wiring in
// cmd/app/main.go initServices (10 nodes, 14 edges).
//
// Note: the Go type is TopologyGraph because Go shares one namespace for types
// and functions; Topology() is the accessor.
type TopologyGraph struct {
	Nodes []Node `json:"nodes"`
	Edges []Edge `json:"edges"`
}

// Node is a single component in the topology graph.
type Node struct {
	ID    string `json:"id"`
	Label string `json:"label"`
	Type  string `json:"type"` // "service" or "db"
}

// Edge is a directed channel/event edge between two nodes.
type Edge struct {
	From  string `json:"from"`
	To    string `json:"to"`
	Label string `json:"label"`
}

// Topology returns the static NMS component graph. Node IDs are the exact
// contract the exporter uses for ComponentIDs.
//
// The graph replicates the ARCHITECTURE.md mermaid diagram: API, EntityService,
// Scheduler, Poller, PluginWorkerPool, MetricsService, PostgreSQL, DiscoveryService,
// DiscoveryWorkerPool, HealthMonitor. Edges map to the channel wiring in
// cmd/app/main.go initServices (crudRequestChan, provisioningEventChan,
// metricRequestChan, deviceChan, discProfileChan, discResultChan,
// schedulerToPollerChan, failureChan, pollResultChan, OpDeactivateDevice) plus
// the discovery pool's Jobs/Results hops.
func Topology() TopologyGraph {
	return TopologyGraph{
		Nodes: []Node{
			{ID: "api", Label: "API", Type: "service"},
			{ID: "entity", Label: "EntityService", Type: "service"},
			{ID: "scheduler", Label: "Scheduler", Type: "service"},
			{ID: "poller", Label: "Poller", Type: "service"},
			{ID: "pluginpool", Label: "PluginWorkerPool", Type: "service"},
			{ID: "metrics", Label: "MetricsService", Type: "service"},
			{ID: "db", Label: "PostgreSQL", Type: "db"},
			{ID: "discovery", Label: "DiscoveryService", Type: "service"},
			{ID: "discoverypool", Label: "DiscoveryWorkerPool", Type: "service"},
			{ID: "health", Label: "HealthMonitor", Type: "service"},
		},
		Edges: []Edge{
			// Mirrors ARCHITECTURE.md mermaid exactly (10 nodes, 14 edges).
			{From: "api", To: "entity", Label: "Request/Reply"},
			{From: "api", To: "metrics", Label: "Request/Reply"},
			{From: "entity", To: "db", Label: "sqlx"},
			{From: "entity", To: "scheduler", Label: "Events"},
			{From: "entity", To: "discovery", Label: "Events"},
			{From: "scheduler", To: "poller", Label: "Devices"},
			{From: "scheduler", To: "health", Label: "Failures"},
			{From: "poller", To: "pluginpool", Label: "Jobs"},
			{From: "pluginpool", To: "metrics", Label: "Results"},
			{From: "pluginpool", To: "health", Label: "Results"},
			{From: "metrics", To: "db", Label: "pgx.CopyFrom"},
			{From: "health", To: "entity", Label: "OpDeactivateDevice"},
			{From: "discovery", To: "discoverypool", Label: "Jobs"},
			{From: "discoverypool", To: "entity", Label: "Results"},
		},
	}
}

// isNodeID reports whether id is a valid topology node id. The exporter uses
// it to derive ComponentIDs so only real components reach the dashboard.
func isNodeID(id string) bool {
	switch id {
	case "api", "entity", "scheduler", "poller", "pluginpool", "metrics", "db", "discovery", "discoverypool", "health":
		return true
	default:
		return false
	}
}

File: crates/nms/src/tracex/topology.rs
//! Static component graph served to the dashboard (Go tracex/topology.go).
//! Mirrors ARCHITECTURE.md mermaid and the channel wiring in main.rs.

use serde::Serialize;

/// A single component in the topology graph.
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct Node {
    pub id: String,
    pub label: String,
    /// "service" or "db".
    pub node_type: String,
}

/// A directed channel/event edge between two nodes.
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct Edge {
    pub from: String,
    pub to: String,
    pub label: String,
}

/// Static component graph (10 nodes, 14 edges — exact replica of
/// ARCHITECTURE.md).
#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct TopologyGraph {
    pub nodes: Vec<Node>,
    pub edges: Vec<Edge>,
}

/// The static NMS component graph. Node IDs are the exact contract the
/// exporter uses for component_ids.
pub fn topology() -> TopologyGraph {
    TopologyGraph {
        nodes: vec![
            Node { id: "api".into(), label: "API".into(), node_type: "service".into() },
            Node { id: "entity".into(), label: "EntityService".into(), node_type: "service".into() },
            Node { id: "scheduler".into(), label: "Scheduler".into(), node_type: "service".into() },
            Node { id: "poller".into(), label: "Poller".into(), node_type: "service".into() },
            Node { id: "pluginpool".into(), label: "PluginWorkerPool".into(), node_type: "service".into() },
            Node { id: "metrics".into(), label: "MetricsService".into(), node_type: "service".into() },
            Node { id: "db".into(), label: "PostgreSQL".into(), node_type: "db".into() },
            Node { id: "discovery".into(), label: "DiscoveryService".into(), node_type: "service".into() },
            Node { id: "discoverypool".into(), label: "DiscoveryWorkerPool".into(), node_type: "service".into() },
            Node { id: "health".into(), label: "HealthMonitor".into(), node_type: "service".into() },
        ],
        edges: vec![
            Edge { from: "api".into(), to: "entity".into(), label: "Request/Reply".into() },
            Edge { from: "api".into(), to: "metrics".into(), label: "Request/Reply".into() },
            Edge { from: "entity".into(), to: "db".into(), label: "sqlx".into() },
            Edge { from: "entity".into(), to: "scheduler".into(), label: "Events".into() },
            Edge { from: "entity".into(), to: "discovery".into(), label: "Events".into() },
            Edge { from: "scheduler".into(), to: "poller".into(), label: "Devices".into() },
            Edge { from: "scheduler".into(), to: "health".into(), label: "Failures".into() },
            Edge { from: "poller".into(), to: "pluginpool".into(), label: "Jobs".into() },
            Edge { from: "pluginpool".into(), to: "metrics".into(), label: "Results".into() },
            Edge { from: "pluginpool".into(), to: "health".into(), label: "Results".into() },
            Edge { from: "metrics".into(), to: "db".into(), label: "pgx.CopyFrom".into() },
            Edge { from: "health".into(), to: "entity".into(), label: "OpDeactivateDevice".into() },
            Edge { from: "discovery".into(), to: "discoverypool".into(), label: "Jobs".into() },
            Edge { from: "discoverypool".into(), to: "entity".into(), label: "Results".into() },
        ],
    }
}

/// Whether id is a valid topology node id. The exporter uses it to derive
/// component_ids so only real components reach the dashboard.
pub fn is_node_id(id: &str) -> bool {
    matches!(
        id,
        "api" | "entity" | "scheduler" | "poller" | "pluginpool" | "metrics" | "db" | "discovery" | "discoverypool" | "health"
    )
}
